rust/src/dataset/scanner.rs

```diff
@@ -40,79 +52,118 @@ pub struct Scanner<'a> {
-            batch_id: 0,
-            reader: None,
-        })
+        }
     }
 
+    /// Projection.
+    ///
+    /// Only seelect the specific columns. If not specifid, all columns will be scanned.
     pub fn project(&mut self, columns: &[&str]) -> Result<&mut Self> {
         self.projections = self.dataset.schema().project(columns)?;
         Ok(self)
     }
 
+    /// Set limit and offset.
     pub fn limit(&mut self, limit: i64, offset: Option<i64>) -> &mut Self {
         self.limit = Some(limit);
         self.offset = offset;
         self
     }
 
-    pub async fn next_batch(&mut self) -> Option<Result<RecordBatch>> {
-        if self.fragment_idx >= self.fragments.len() {
-            return None;
-        }
+    /// The schema of the output, a.k.a, projection schema.
+    pub fn schema(&self) -> SchemaRef {
+        Arc::new(ArrowSchema::from(&self.projections))
+    }
 
-        let fragment = &self.fragments[self.fragment_idx];
-        // Only support 1 data file (no schema evolution for now);
-        assert!(fragment.files.len() == 1);
-        let data_file = &fragment.files[0];
-        let path = self.dataset.data_dir().child(data_file.path.clone());
-        if self.reader.is_none() {
-            self.reader = Some(
-                FileReader::new(
-                    &self.dataset.object_store,
-                    &path,
-                    Some(&self.dataset.manifest),
-                )
-                .await
-                .unwrap(),
-            );
-            self.reader
-                .as_mut()
-                .map(|reader| reader.set_projection(self.projections.clone()));
-        }
+    /// Create a stream of this Scanner.
+    ///
+    /// TODO: implement as IntoStream/IntoIterator.
+    pub fn into_stream(&self) -> ScannerStream {
+        const PREFECTH_SIZE: usize = 8;
+        let object_store = self.dataset.object_store.clone();
+
+        let data_dir = self.dataset.data_dir().clone();
+        let fragments = self.fragments.clone();
+        let manifest = self.dataset.manifest.clone();
+
+        ScannerStream::new(
+            object_store,
+            data_dir,
+            fragments,
+            manifest,
+            PREFECTH_SIZE,
+            &self.projections,
+        )
+    }
+}
 
-        if let Some(reader) = &self.reader {
-            let batch = reader.read_batch(self.batch_id).await;
-            self.batch_id += 1;
-            if self.batch_id as usize >= reader.num_batches() {
-                self.batch_id = 0;
-                self.fragment_idx += 1;
-                self.reader = None;
+pub struct ScannerStream {
+    rx: Receiver<Result<RecordBatch>>,
+}
+
+impl ScannerStream {
+    fn new(
+        object_store: Arc<ObjectStore>,
+        data_dir: Path,
+        fragments: Vec<Fragment>,
+        manifest: Arc<Manifest>,
+        prefetch_size: usize,
+        schema: &Schema,
+    ) -> Self {
+        let (tx, rx) = mpsc::channel(prefetch_size);
```

Comment:
Let's confirm this doesn't make extra copies first? Otherwise we'd have to rip it back out?

Reply:
channel does copy, but `RecordBatch` itself contains the pointer to schema / array.

https://docs.rs/arrow-array/30.0.1/src/arrow_array/record_batch.rs.html#55-56

Guess the main overhead would be copy a vector of points to the arrays. 

```rust
pub struct RecordBatch {
    schema: SchemaRef,
    columns: Vec<Arc<dyn Array>>,  <---- Here
   ..
}
```

Let's benchmark this later, if the copy really is a concern, we can just pass `Arc<RecordBatch>` into the channel.